### PR TITLE
Coap library support over sockets

### PR DESCRIPTION
**Pull request by rveerama1**
CoAP library is migrated to support over socket based
applications or other higher layer protocols. Most of the
API's and functionality is kept as it is except few changes.

net_pkt/net_buf is removed from CoAP library. Now it expects
a pre-allocated flat buffer and length. If there is not enough
space to append any data, library simply returns an error.
It's user's responsibility to allocate and free memory.

One change in functionality is, earlier coap_pending_clear()
used to clear the memory, but now it's user's responsibility
to free the memory.

Samples and tests improved a bit.

### CODE REVIEW

**Comment by mike-scott** [comment]
Hi @rveerama1 
I see that we do indeed have similar approaches between this and https://github.com/zephyrproject-rtos/zephyr/pull/10110

A couple of quick notes:
- Can we decide on an internal set of shared "append / insert / read" functions for standard buffers?  I will need them for LwM2M as well and I'm sure future users will too.
- Do we really need to copy the entire library?  If we migrate existing net_app users (similar to what I did  in my PR) this seems unnecessary.
- If we do in fact, need to make a copy, does it need to say "sock" in the name?  It's actually network layer agnostic at this point.

**Comment by rveerama1** [comment]
>     * Can we decide on an internal set of shared "append / insert / read" functions for standard buffers?  I will need them for LwM2M as well and I'm sure future users will too.

Sure, I agree that.Generic helpers on set of apis like "read, write, append and insert" definitely useful for users of standard buffers.

**Comment by rveerama1** [comment]
>     * Do we really need to copy the entire library?  If we migrate existing net_app users (similar to what I did  in my PR) this seems unnecessary.

We never know who else is depending on this CoAP other than part of upstream Zephyr. IMO it's not good practice to wipe out entire library with new migrated apis in a single release. Let's keep original as it is for this release and make sure, all users of upstream Zephyr migrate to new one first and then remove old CoAP immediately after release.

**Comment by rveerama1** [comment]
>     * If we do in fact, need to make a copy, does it need to say "sock" in the name?  It's actually network layer agnostic at this point.

That's true, "coap_sock" was not a good name. But you know I pushed my wip branch just because you are flooding with different versions :) (it's a compliment though), may be because you have some deadlines to finish lwm2m over sockets?.

I am taking some time finish this migration just because of improving coap-client and tests. coap-server sample covers most of the functionality, but client I always depends on something else. It was like that from ZOAP era.

Overall, to avoid duplication of our work, how do we proceed?

**Comment by mike-scott** [comment]
> That's true, "coap_sock" was not a good name. But you know I pushed my wip branch just because you are flooding with different versions :) (it's a compliment though), may be because you have some deadlines to finish lwm2m over sockets?.
>
> I am taking some time finish this migration just because of improving coap-client and tests. coap-server sample covers most of the functionality, but client I always depends on something else. It was like that from ZOAP era.
>
> Overall, to avoid duplication of our work, how do we proceed?

@rveerama1 please finish your migration for CoAP.  I will stop flooding versions.  Your design is close enough to mine, that I can proceed now, without my version.

**Comment by rveerama1** [comment]
> @rveerama1 please finish your migration for CoAP. I will stop flooding versions. Your design is close enough to mine, that I can proceed now, without my version.

I am almost done with CoAP related changes. coap_server sample should provides all the functionality like current one. I am improving coap_client sample. I tested coap_server with "libcoap" / etsi_testcases.sh. Working like previous one. I haven't modified tests yet.

In mean time can you please start working on top of this PR and comment. I will continue on coap_client and tests. Thanks.

File: samples/net/sockets/coap_client/README.rst [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+.. code-block:: none
+
+    coap://[ff02::fd]:5683/test
+
+Building and Running
+********************
+
+This project will print all the octets of the response received, more context can
+be obtained by using a tool like tcpdump or wireshark.
```
Comment (dbkinder) [inline samples/net/sockets/coap_client/README.rst]: change ``like`` to ``such as``

File: samples/net/sockets/coap_client/README.rst [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+See the `net-tools`_ project for more details.
+
+It can be built and executed on QEMU as described in :ref:`networking_with_qemu`.
```
Comment (dbkinder) [inline samples/net/sockets/coap_client/README.rst]: change ``It`` to ``This sample``

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: subsys/net/lib/coap_sock/coap_sock.c [hunk 3]
```diff
@@ -0,0 +1,1402 @@
...
+	for (i = 0, p = pendings; i < len; i++, p++) {
+		if (!p->timeout) {
+			continue;
+		}
+
+		if (resp_id != p->id) {
+			continue;
+		}
+
+		return p;
```
Comment (mike-scott) [inline subsys/net/lib/coap_sock/coap_sock.c]: Is there a reason why the "coap_pending_clear(p);" was removed here?

**Review by mike-scott — CHANGES_REQUESTED** [review]

**Comment by rveerama1** [comment]
> @rveerama1 How did you want to share a standard set of buffer handling functions? (IE: what I called fbuf in my current PR)

@mike-scott At-least in this version don't want tie up with any net_buf/flat_buf related functions. Simple u8_t *data pointer and length.

Sure higher layer protocols or apps over sockets need those kind of buffer mechanism and helper functions to deal with buffers. It should be generic and can be introduced with separate PR. I didn't want to mix these PRs.

File: subsys/net/lib/coap_sock/coap_sock.c (hunk 3 above)
Comment (rveerama1) [reply subsys/net/lib/coap_sock/coap_sock.c]: Yes, CoAP lib doesn't how that buffer is allocated. So it's caller responsibility to free that.
I changed in samples(coap_server) like:

k_free(pending->data);
coap_pending_clear(pending);

**Comment by rveerama1** [comment]
@jukkar @mike-scott @dbkinder PR ready for review. Please take a look.

**Comment by mike-scott** [comment]
@rveerama1 I will re-test this PR later today.  Sorry for the delay.

**Comment by rveerama1** [comment]
Shippable complaints about Duplicate ID in CoAP documentation. Upstream CoAP has @defgroup "coap" and this PR has @defgroup "coap_sock". These ID's are different, But still I don't understand why it complaints. To make it easy for migration of CoAP based applications and other higher layer protocols, I almost kept API's and structures names as it is. I expected different @defgroup name will help. But it's not. I don't know how exactly solve this issue.

@dbkinder do you have any idea why?

**Comment by mike-scott** [comment]
@rveerama1 I tested this PR w/ my latest work on LwM2M sockets and it seems to be working, with 1 really interesting exception.  I'm getting double logger lines occasionally:
```
uart:~$ ***** Booting Zephyr OS 1.13.99 *****
[00:00:00.231,326] <dbg> net_lwm2m_engine: LWM2M engine socket receive thread started
[00:00:00.235,998] <dbg> net_lwm2m_engine: LWM2M engine periodic work started
[00:00:00.238,064] <dbg> net_lwm2m_obj_security: Create LWM2M security instance: 0
[00:00:00.240,043] <dbg> net_lwm2m_obj_server: Create LWM2M server instance: 0
[00:00:00.241,868] <dbg> net_lwm2m_obj_device: Create LWM2M device instance: 0
[00:00:00.243,328] <dbg> net_lwm2m_obj_firmware: Create LWM2M firmware instance: 0
[00:00:12.723,008] <inf> net_lwm2m_client_app: Run LWM2M client
[00:00:12.723,008] <inf> net_lwm2m_client_app: Run LWM2M client
uart:~$ [00:00:12.741,342] <dbg> net_lwm2m_engine: path:0/0/2, value:0x0041b398, len:1
[00:00:12.748,577] <dbg> net_lwm2m_engine: path:3/0/9, value:0x0041b398, len:1
[00:00:12.751,646] <dbg> net_lwm2m_engine: path:3/0/10, value:0x0041b3a8, len:4
[00:00:12.755,082] <dbg> net_lwm2m_engine: path:3/0/20, value:0x0041b398, len:1
[00:00:12.758,251] <dbg> net_lwm2m_engine: path:3/0/21, value:0x0041b3a8, len:4
[00:00:12.763,676] <dbg> net_lwm2m_engine: path:3303/0
[00:00:12.765,576] <dbg> net_ipso_temp_sensor: Create IPSO Temperature Sensor instance: 0
[00:00:12.767,650] <dbg> net_lwm2m_engine: path:3303/0/5700, value:0x0041b3b4, len:8
[00:00:12.772,738] <inf> net_lwm2m_rd_client: LWM2M Client: qemu_x86
[00:00:12.772,738] <inf> net_lwm2m_rd_client: LWM2M Client: qemu_x86
uart:~$ [00:00:20.362,996] <dbg> net_lwm2m_engine: path:0/0/1, buf:0x00421f77, buflen:1
[00:00:20.366,103] <dbg> net_lwm2m_engine: path:1/0/1, buf:0x004049a4, buflen:4
[00:00:20.368,558] <dbg> net_lwm2m_rd_client: sec_obj_inst: 0
[00:00:20.370,670] <inf> net_lwm2m_rd_client: RD Client started with endpoint 'qemu_x86' with client lifetime 30
[00:00:20.375,134] <dbg> net_lwm2m_engine: URL: coap://2001:db8::2
[00:00:20.409,541] <dbg> net_lwm2m_rd_client: registration sent
[00:00:20.370,670] <inf> net_lwm2m_rd_client: RD Client started with endpoint 'qemu_x86' with client lifetime 30
uart:~$ [00:00:20.518,919] <dbg> net_lwm2m_engine: checking for reply from [2001:db8::2]
[00:00:20.526,362] <dbg> net_lwm2m_rd_client: Registration callback (code:2.1)
[00:00:20.528,181] <dbg> net_lwm2m_client_app: Registration complete
[00:00:20.529,386] <inf> net_lwm2m_rd_client: Registration Done (EP='nPtXRyXZdg')
[00:00:20.531,417] <dbg> net_lwm2m_engine: reply 0x00400178 handled and removed
[00:00:20.529,386] <inf> net_lwm2m_rd_client: Registration Done (EP='nPtXRyXZdg')
uart:~$ [00:01:57.339,760] <dbg> net_lwm2m_rd_client: registration sent
[00:01:57.370,150] <dbg> net_lwm2m_engine: checking for reply from [2001:db8::2]
[00:01:57.371,956] <inf> net_lwm2m_rd_client: Update callback (code:2.4)
[00:01:57.373,067] <dbg> net_lwm2m_client_app: Registration update complete
[00:01:57.374,085] <inf> net_lwm2m_rd_client: Update Done
[00:01:57.375,148] <dbg> net_lwm2m_engine: reply 0x00400178 handled and removed
[00:01:57.371,956] <inf> net_lwm2m_rd_client: Update callback (code:2.4)
[00:01:57.374,085] <inf> net_lwm2m_rd_client: Update Done
uart:~$ [00:00:42.486,684] <dbg> net_lwm2m_rd_client: registration sent
[00:00:42.514,336] <dbg> net_lwm2m_engine: checking for reply from [2001:db8::2]
[00:00:42.517,287] <inf> net_lwm2m_rd_client: Update callback (code:2.4)
[00:00:42.519,260] <dbg> net_lwm2m_client_app: Registration update complete
[00:00:42.521,200] <inf> net_lwm2m_rd_client: Update Done
[00:00:42.522,990] <dbg> net_lwm2m_engine: reply 0x00400178 handled and removed
[00:00:42.517,287] <inf> net_lwm2m_rd_client: Update callback (code:2.4)
[00:00:42.521,200] <inf> net_lwm2m_rd_client: Update Done
```

That wasn't happening the last time I tested.. any idea where that's coming from?

**Comment by mike-scott** [comment]
I should say, I tested these patches + my patches on today's master branch.

**Comment by rveerama1** [comment]
> I should say, I tested these patches + my patches on today's master branch.

Ok, thank you.

**Comment by dbkinder** [comment]
@rveerama1 As for the documentation errors, the sphinx/breath system appears to be complaining about duplication definitions because, well, there are many duplicate definitions.  The two defgroups (coap and coap_sock) have many common declarations for types and functions.  I suspect in an application build, you would pick one of these headers to include.  But for the documentation build, we see both of the header files and the cross-referencing system doesn't like all these duplicates.  (If you included both header files in your app build I suspect the compiler wouldn't be happy either.)  For this to work with our documentation build, assuming these are alternate implementations of the same interface, we'll need to exclude one of them; easily done in the doxygen configuration file.   

There's more discussion of this in https://github.com/sphinx-doc/sphinx/issues/4820 (and other sphinx issues linked to there). 

Any of you developers have any other insights to add?

**Comment by jukkar** [comment]
sanitycheck error will be fixed by #10763

**Comment by mike-scott** [comment]
@dbkinder I will probably have the same issues with the LwM2M socket changes here: https://github.com/zephyrproject-rtos/zephyr/pull/9832 where I have both lwm2m and lwm2m_sock.  I'll follow this conversation to see what the decision is.

**Comment by rveerama1** [comment]
@mike-scott found a way to exclude these and avoid errors. https://github.com/zephyrproject-rtos/zephyr/pull/10121/commits/d02884703741602ab97ffa83b1f486ad290212a8#diff-d68d5906013fd46e734966426f6a274e

**Comment by mike-scott** [comment]
@rveerama1 ah very nice.  thanks

**Comment by rveerama1** [comment]
That was a trial and error. Now I pushed latest version which excludes coap and generates documentation for coap_sock.

**Comment by rveerama1** [comment]
@dbkinder thank you, now sanitycheck is fine, please take a look.

File: doc/zephyr.doxyfile.in [hunk 4]
```diff
@@ -759,7 +759,7 @@ INPUT                  = @ZEPHYR_BASE@/include/ \
                          @ZEPHYR_BASE@/include/arch/arm/cortex_m \
                          @ZEPHYR_BASE@/include/arch/nios2/ \
                          @ZEPHYR_BASE@/lib/libc/minimal/include/ \
-                         @ZEPHYR_BASE@/include/net/coap.h \
+                         @ZEPHYR_BASE@/include/net/coap_sock.h \
```
Comment (dbkinder) [inline doc/zephyr.doxyfile.in]: Technically, since ``RECURSIVE = YES`` is specified, all the subdirectories of ``/include`` are scanned so we don't need to have any of these directories or files specified within the ``/include`` folder, so this change isn't needed.  (I should submit a separate PR to remove all these folder/file names that would be included by mentioning ``/include``.)

File: doc/zephyr.doxyfile.in [hunk 5]
```diff
@@ -807,7 +807,9 @@ RECURSIVE              = YES
 EXCLUDE                = @ZEPHYR_BASE@/include/spi_legacy.h \
                          @ZEPHYR_BASE@/include/net/http_legacy.h \
                          @ZEPHYR_BASE@/include/misc/util.h \
-                         @ZEPHYR_BASE@/include/cmsis_rtos_v1/cmsis_os.h
+                         @ZEPHYR_BASE@/include/cmsis_rtos_v1/cmsis_os.h \
+                         @ZEPHYR_BASE@/include/net/coap.h \
```
Comment (dbkinder) [inline doc/zephyr.doxyfile.in]: This is the real "trick" for fixing the duplicate names errors (since coap.h and coap_sock.h define conflicting names).

**Review by dbkinder — APPROVED** [review]
+1 for doc changes (and doxygen config change)

File: doc/zephyr.doxyfile.in (hunk 4 above)
Comment (rveerama1) [reply doc/zephyr.doxyfile.in]: > I should submit a separate PR to remove all these folder/file names that would be included by mentioning `/include`

ok

**Comment by rveerama1** [comment]
> +1 for doc changes (and doxygen config change)

Thank you.

**Comment by rveerama1** [comment]
@mike-scott can you update your review or let me know if there is anything to be done.

**Comment by rveerama1** [comment]
@mike-scott ping

**Comment by mike-scott** [comment]
@rveerama1 apologies for my delay.  I'm leaving the Scotland area (post ELC work).

I expect to re-run my tests with this PR in the next 24-36 hours or so.

**Comment by rveerama1** [comment]
@mike-scott sorry, I didn't know that you are at ELC. Take your time.

**Comment by mike-scott** [comment]
@rveerama1 Tested today on master w/ latest patches and this PR LGTM.    If I end up finding other issues we can address those in a new PR.

**Review by mike-scott — APPROVED** [review]

**Review by nashif — APPROVED** [review]
